lektor/db.py

```diff
@@ -281,6 +287,30 @@ def __eval__(self, record):
         return self.__value
 
 
+class _SeqExpr(Expression):
+    def __init__(self, value):
+        try:
+            iter(value)
+        except TypeError:
+            value = [value]
+        self.__evaluated = False
+        self.__value = value
+
+    def __eval__(self, record):
+        def eval(item):
+            if isinstance(item, Record):
+                return item["_id"]
+            if isinstance(item, Expression):
+                return item.__eval__(record)
+            return item
+
+        if self.__evaluated:
+            return self.__value
+        self.__value = [eval(v) for v in self.__value]
```

Comment:
I think this caching is not correct.  What happens if `__eval__` is called a second time with a different `record` (and the sequence contains an `Expression`)? 

(Is the caching even worth the effort?  The evaluation of non-`Expression` values is cheap.  The `Expression`-valued items are potentially heavy, but how often do they get re-evaluated for the same `record`?)

----

Cache or not, there is another optimization that could be made here.  I'm not completely sure it's worth the effort, but it's worth considering.

Instead of always evaluating every item in the sequence, `__eval__` could return an iterator/generator that evaluates the items one at a time, as they are iterated over, thus saving some (potentially heavy) evaluations if the calling `is in` operator hits early.

This would make the caching logic more complicated, but not overly so.
If the caching is eliminated, then this optimization is simple enough that it's almost certainly worth making.

Reply:
Yes the caching is a worry. I'll remove the `__evaluated` test.

But returning a generator looks problematic too. Since the _ContainmentExpr into which
it is placed (see the `isin` method) might be evalued more that once in which case the generator will be already
exhausted on the second evaluation.

Reply:
Oops! I take the second paragraph back. _ContainmentExpr re-evalues seq for each on its __eval__'s

Reply:
OK. Made the evaluation "lazier" but still had to guard against multiple evaluations of a generator

Reply:
This won't pick up the case for value being a generator (which can be the case -- my test in 
test_db.py fail). I could use isinstance(value, Iterable) *but*....

There is a footnote in https://docs.python.org/3/library/collections.abc.html that says:

Checking isinstance(obj, Iterable) detects classes that are registered as Iterable or that have an __iter__() method, but it does not detect classes that iterate with the __getitem__() method. The only reliable way to determine whether an object is [iterable] is to call iter(obj)

Which the reason for the iter(value) test. 

Also... I don't know if I should test for value being a string. Since currently: F.name.isin("ABC") == F.name.isin(['A','B','C'])
(since list("ABC") => ['A','B','C'], which is "probably" not what a user intended.

Reply:
> This won't pick up the case for value being a generator

I'm pretty sure it will.  Note that the `if isinstance(value, Collection)` check is just an optimization — it could be omitted altogether.  The check for iterability happens at `try: self.__value = list(value)` (which is essentially the same test as your `try: iter(value)`, except that now we pre-cache the results of the iteration.

The idea behind the `if instance(value, Collection)` check is to identify those values for which `iter(value)` always returns a "fresh" iterator.  Since we can already iterate these values multiple times without issue, we can skip the `value = list(value)` cache step.

For the sake of concreteness, here is my previous proposal, in its entirety.  I've rewritten the constructor for (I think) clarity, but it's equivalent to my previous proposal.)
```py
class _SeqExpr(Expression):
    def __init__(self, value):
        if not isinstance(value, collections.abc.Collection):
            try:
                value = list(value)
            except TypeError:
                value = [value]
        self.__value = value

    def __eval__(self, record):
        for item in self.__value:
            if instance(item, record):
                yield item["_id"]
            elif isinstance(item, Expression):
                yield item.__eval__(record)
            else:
                yield item
```
Since `Collection`s are iterable and sized, the assumption made here is that `iter(collection)` always produces a fresh iterable (since one would expect `len(list(iter(collection))) == len(collection)` to always be true.)

----

> Also... I don't know if I should test for value being a string. Since currently: F.name.isin("ABC") == F.name.isin(['A','B','C']) (since list("ABC") => ['A','B','C'], which is "probably" not what a user intended.

I don't have strong feelings, but, probably should either go all-in on the [DWIM](https://en.wikipedia.org/wiki/DWIM) and test for string, or skip the DWIM altogether and just throw TypeError if `value` is not iterable.  I can see arguments either way.  My personal feeling (at the moment) is explicit is better, and skip the DWIM.  But, since `Expression`s are largely intended for use in templates, I can see where the DWIM approach is reasonable, too.

If one does go full DWIM, perhaps things should be made so that `F.name.isin('Fred', 'Joe')` is equivalent to  `F.name.isin(['Fred', 'Joe'])`?  In that case the DWIM logic should probably be moved to `Expression.isin` and `Expression.isnotin` (abstracted to a separate helper function).